README.md
# RDS Python
## WARNING: THIS PROJECT IS IN EARLY DEVELOPMENT STAGE. CONTENT OR CODE SHOULD ONLY BE USED FOR TESTING OR EVALUATION PURPOSES.
[![Build Status](https://travis-ci.com/mtna/rds-python.svg?branch=master)](https://travis-ci.org/mtna/rds-python) 
[![Coverage Status](https://coveralls.io/repos/github/mtna/rds-python/badge.svg?branch=master&service=github)](https://coveralls.io/github/mtna/rds-python?branch=master)
[![PyPI version](https://badge.fury.io/py/mtna-rds.svg)](https://badge.fury.io/py/mtna-rds)
![Python Version](https://img.shields.io/badge/python-2.7|3.6|3.7|3.8-blue)  
[![License](https://img.shields.io/badge/license-apache_2.0-green)](https://www.apache.org/licenses/LICENSE-2.0)
[![Code Style](https://img.shields.io/badge/code_style-black-black)](https://pypi.org/project/black/)
  
This python module utilizes MTNA's Rich Data Services API to quickly and efficiently access data sets and metadata stored in our repository. Through RDS, you can easily perform complex queries and tabulations on the data you are interested in while also getting back any relevant metadata.

RDS greatly simplifies the long process finding the data to begin with, cleaning and transforming the data, and converting the data into a dataframe. All of this is done in a single step using our queries. This lets you focus on the analyzing and visualizing of the data instead of managing it.  

## References
[RDS API Documentation](https://covid19.richdataservices.com/rds/swagger/) | [Examples](https://github.com/mtna/rds-python-examples) | [Contributing](CONTRIBUTING.md) | [Developer Documentation](DEVELOPER.md) | [Changelog](CHANGELOG.md)
|---|---|---|---|---|

**Contents:**  
- [Announcements](#announcements)  
- [Installation](#installation)  
- [Usage](#usage)  
- [About](#about)  
- [Software](#software)  
- [License](#license)  
    
## Announcements
### Version v0.2.18 released
This version of **RDS** Python has added the feature of using an API key when connecting to an instance of RDS.

## Installation
### Using pip
Use the package manager [pip](https://pip.pypa.io/en/stable/) to install rds python
```bash
pip install mtna-rds
```

## Usage
Through the **RDS** API, you care able to query for records of data as well as perform a tabulation. Both a simple query and a tabulation contain options for grouping, ordering and filtering of the data, as well as specifying if metadata is wanted or not.

The data returned by a query/tabulation will be contained within an `RdsResults` object. This object has three properties: one is the records of data that can be used to build out a dataframe for a graph or chart, one is the column names for each column of data in the records, and the last is a collection of metadata in JSON format that provides information that can be used for better analyzation of your data.

### Select Query
Imagine that you would like to get some demographic data in the United State. You look through our **Catalog** and see that we have the data you are interested in. The first thing you would need to do to access this data is to establish a link to the demographic dataset that we host in our repository. To do this, you simply create a `DataProduct` with the **ID** of the dataproduct that contains the demographic information and the **ID** of the catalog that contains the dataproduct.
```python
from rds import Server

server = Server("domain", "api_key")
catalog = server.get_catalog("catalog_id")
dataproduct = catalog.get_dataproduct("dataproduct_id")
```

Once the `DataProduct` is created, you can perform your query and get back the results (which contains records in a dataframe). If you wanted to know how many people were born between the years 1900 and 1950 for each year, you could perform the following query.
```python
results = dataproduct.select(cols=["year_of_birth", "amount_born:count(*)"], where=["year_of_birth>1900"], orderby=["year_of_birth"], groupby=["year_of_birth"], limit=50)
```

This query tells **RDS** that we want the year of birth for each records as well as the number of records with that year of birth (where we are renaming the column to "amount_born"). We then filter for everyone born after 1900. We also make sure the data is in the correct order and then group the data by year of birth so that we only have a single record returned per year. Setting the limit to 50 ensures we only get date from years 1900 to 1950 (assuming there are no missing years of data).

After obtaining the data, you can pull out the records and columns and place directly into a dataframe for use in a graph or chart. Below we demonstrate by building out a simple line plot of people born per year, utilizing the pandas package.
```python
import pandas as pd

dataframe = pd.DataFrame(results.records, columns=results.columns)

sns.lineplot(data=dataframe, x=dataframe.columns[0], y=dataframe.columns[1])
plt.show()
```

### Tabulation Query
A tabulation query is used almost identically to a select query, except it uses different parameters as a tabulation is more useful for checking the relationships between columns of data

If you wanted to know the amount of male/females for each race in the census, you would perform the below tabulation query.
```python
results = dataproduct.tabulate(dims=["sex", "race"], measure=["count(*)"], orderby=["race"], inject=True)
```

You can think of the parameter `dims` as the dimension of a tabulation table, and the parameter `measure` as the value that you want in each cell of the table. One thing you may notice that is new is the `inject` parameter. This signifies that we want to replace any "coded" values with their more readable labels. Sex can be an example of a "coded" value as many times the data is coded as "1" to refer to male and a "2" to refer to female. Since "1" and "2" would not be very descriptive in a chart, **RDS** gives you the ability to replace them with what the codes actually mean.

### Metadata
Metadata can be directly asked for on any of our resources. This includes the server, catalogs, dataproducts, variables, classifications, and codes. The metadata contains extensive information on what the resource is and what it is used for.

## About
This project is developed and maintained by [MTNA](https://www.mtna.us/).

More detailed documentation about what the current version of RDS can do can be found [here](https://documenter.getpostman.com/view/2220438/SzS4QmXD?version=latest#intro.)

If you are interested in using the RDS framework directly, you can visit our site [here](https://www2.richdataservices.com/).

## Software
Compatible with Python 2.7 and Python 3.6 and higher.

If using python 3, it is recommended that you utilize [pandas](https://pandas.pydata.org/) dataframes when working with any records returned from an RDS query.

The are no dependencies required to run RDS Python.

## License
[Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)


CHANGELOG.md
# v0.2.18 (2022-7-1)
## Added
- `api_key` parameter to Server class, allows connecting to secure RDS host.

# v0.2.16 (2020-7-16)
## Changed
- limit defaults to `None` instead of 20

# v0.2.15 (2020-7-16)
Parameters modified and new ones added. Bugs fixed and tests created.
## Added
- `rds_format` parameter to select and tabulate queries, changes json output format
- `count` parameter to select and tabulate queries, returns row count
- `groupby` parameter to tabulate query
- returned RdsResults now has `count` and `totals` attributes
## Changed
- `include_metadata` parameter changed to `metadata`
- `limit` parameter default changed from 1000 to 20
- `measure` tabulate parameter default changed from count(*) to None
## Fixed
- query batching accounting for 10000 cell limit as opposed to row limit
## Removed
- batching queries in a tabulate query

# v0.2.14 (2020-7-08)
Modifying query parameters.
## Added
- `weights` parameter added to `select` and `tabulate` queries
## Fixed
- `inject_metadata` parameter changed to `include_metadata`

# v0.2.13 (2020-7-02)
Added additional Select parameter.
## Added
- Added `inject` parameter to inject labels into code values returned

# v0.2.12 (2020-6-30)
Bug fix.
## Fixed
- Fixed bug appending column names in results returned

# v0.2.11 (2020-6-30)
## Added
- Better error handling

# v0.2.0 (2020-6-30)
Created new classes to better encapsulate functions
## Added
- Server class for server information
- Catalog class for catalog information

# v0.1.2 (2020-5-21)
Bug fix, added more comprehensive tests.
## Fixed
- Default measure value in `tabulate()`

# v0.1.1 (2020-5-18)
Bug fix.
## Fixed
- Code metadata retrieval

# v0.1.0 (2020-5-15)
Initial commit.
## Added
- Select queries
- Tabulate queries
- Metadata retrieval


rds/dataproduct.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Retrieves data and metadata from RDS.
"""

# Built-in/Generic Imports
import sys
import json
import math

from .utility import get_response, check_valid


#TODO pass api key to util methods
class DataProduct:
    """
    Holds information to connect to a data product and allows methods for querying it.
    
    Parameters
    ----------
    api : str, required
        The url hosting an RDS server
    catalog_id : str, required
        ID of the catalog, required
    dataproduct_id : str, optional
        ID of the data product. Default is None
    """

    def __init__(self, api, api_key, catalog_id, dataproduct_id):
        metadata = check_valid(
            api + "/api/catalog/" + catalog_id + "/" + dataproduct_id,
            api_key, "Invalid dataproduct ID",
            is_json=True,
        )
        self.api = api
        self.api_key = api_key
        self.catalog_id = catalog_id
        self.dataproduct_id = dataproduct_id
        self.name = metadata["name"]
        self.description = metadata["description"]
        self.last_update = metadata["lastUpdate"]
        self.uri = metadata["uri"]

    def count(self):
        """
        Gets the count of records

        Returns
        -------
        int
            The record count.

        """
        api_call = self._get_url("query") + "/count"
        response = get_response(api_call, self.api_key)
        return json.load(response)

    def select(
        self,
        cols=None,
        where=None,
        orderby=None,
        groupby=None,
        collimit=None,
        coloffset=0,
        weights=None,
        metadata=True,
        inject=False,
        count=False,
        limit=None,
        offset=0,
        rds_format=None,
    ):
        """
        Queries the data product for a set of records.

        Parameters
        ----------
        cols : list of str, required
            the columns of the records to be returned. The default is None which queries for all columns.
        where : list of str, optional
            filtering by comparative and conjunctive operators. The default is None.
        orderby : list of str, optional
            orders the records by a one or more columns. The default is None.
        groupby : list of str, optional
            groups the records by one or more columns. The default is None.
        collimit : int, optional
            limit of the columns in the data frame. The default is 1000.
        coloffset : int, optional
            offset of the columns in the data frame. The default is 0.
        weights : list of str, optional
            columns to be weighed. The default is None.
        metadata : bool, optional
            flag for if metadata should be used/returned with the data frame. Setting this to false will also cause no column names to return in the results. The default is True.
        inject : bool, optional
            flag for if the code labels should be used over the code values. The default is False.
        count : bool, optional
            flag for returning the record count in the info. The default is False.
        limit : int, optional
            limit for the records in the data frame. The default is 1000.
        offset : int, optional
            offset for the records in the data frame. The default is 0.
        rds_format : string, optional
            the format of the json object returned. The default is mtna_simple.

        Returns
        -------
        results : object
            A wrapper object for the dataframe and metadata.

        """
        api_call = self._get_url("query") + "/select?"
        params = {}
        params.update(self._get_param(cols, "cols"))
        params.update(self._get_param(where, "where"))
        params.update(self._get_param(orderby, "orderby"))
        params.update(self._get_param(groupby, "groupby"))
        params.update(self._get_param(collimit, "collimit"))
        params.update(self._get_param(coloffset, "coloffset"))
        params.update(self._get_param(weights, "weights"))
        params.update(self._get_param(rds_format, "format"))
        params.update(self._get_param(str(metadata).lower(), "metadata"))
        params.update(self._get_param(str(inject).lower(), "inject"))
        params.update(self._get_param(str(count).lower(), "count"))

        max_records = limit
        col_count = self._get_column_count(cols, collimit)
        if limit == None or limit * col_count > 10000:
            limit = math.floor(10000 / col_count)

        results = self._batch(api_call, params, max_records, limit, offset)

        metadata_json = None
        if metadata:
            metadata_json = _get_metadata(results)

        count_value = None
        if count:
            count_value = results[0]["info"]["rowCount"]

        return _get_rds_results(results, metadata_json, count_value)

    def tabulate(
        self,
        dims=None,
        measure=None,
        where=None,
        orderby=None,
        groupby=None,
        weights=None,
        totals=False,
        metadata=True,
        inject=False,
        count=False,
        rds_format=None,
    ):
        """
        Queries the data product for a set of tabulated records.

        Parameters
        ----------
        dims : list of str, required
            dimensions of the tabulation, represents the rows and table of table. Default is None.
        measure : list of str, optional
            value the tabulation is calculating from the measures, represents the values in the
            cells of a table. The default is count(*).
        where : list of str, optional
            filtering by comparative and conjunctive operators. The default is None.
        orderby : list of str, optional
            orders the records by one or more columns. The default is None.
        groupby : list of str, optional
            groups the records by one or more columns. The default is None.
        weights : list of str, optional
            columns to be weighed. The default is None.
        totals : bool, optional
            flag for if the totals should be returned with the data frame. The default is False.
        metadata : bool, optional
            flag for if metadata should be used/returned with the data frame. The default is True.
        inject : bool, optional
            flag for if the code labels should be used over the code values. The default is False.
        count : bool, optional
            flag for returning the record count in the info. The default is False.
        rds_format : string, optional
            the format of the json object returned. The default is mtna_simple.

        Returns
        -------
        results : object
            A wrapper object for the dataframe and metadata.
        """
        api_call = self._get_url("query") + "/tabulate?"
        params = {}
        params.update(self._get_param(dims, "dims"))
        params.update(self._get_param(measure, "measure"))
        params.update(self._get_param(where, "where"))
        params.update(self._get_param(orderby, "orderby"))
        params.update(self._get_param(groupby, "groupby"))
        params.update(self._get_param(weights, "weights"))
        params.update(self._get_param(rds_format, "format"))
        params.update(self._get_param(str(metadata).lower(), "metadata"))
        params.update(self._get_param(str(inject).lower(), "inject"))
        params.update(self._get_param(str(totals).lower(), "totals"))
        params.update(self._get_param(str(count).lower(), "count"))

        results = [_query(api_call, self.api_key, params)]

        metadata_json = None
        if metadata:
            metadata_json = _get_metadata(results)

        count_value = None
        if count:
            count_value = results[0]["info"]["rowCount"]

        return _get_rds_results(results, metadata_json, count_value)

    def get_variable(self, variable=None):
        """
        Gets the metadata for one or more variables in JSON format.

        Parameters
        ----------
        variable : list of str, optional
            A list of variable names you want the metadata of. The default is None which
            will return metadata for all variables.

        Returns
        -------
        metadata : JSON
            Detailed information surrounding the variable(s).

        """
        api_call = self._get_url("catalog")
        if variable is None:
            api_call += "/variables"
        else:
            api_call += "/variable/" + variable

        response = get_response(api_call, self.api_key)
        return json.load(response)

    def get_classification(self, classification=None):
        """
        Gets the metadata for one or more classifications in JSON format.

        Parameters
        ----------
        classification : list of str, optional
            A list of classification names you want the metadata of. The default is None which
            will return metadata for all classifications.

        Returns
        -------
        metadata : JSON
            Detailed information surrounding the classification(s).

        """
        api_call = self._get_url("catalog")
        if classification is None:
            api_call += "/classifications"
        else:
            api_call += "/classification/" + classification

        response = get_response(api_call, self.api_key)
        return json.load(response)

    def get_code(self, classification, limit=20):
        """
        Gets the metadata for codes in JSON format.

        Parameters
        ----------
        classification : str
            The name of the classification you want the codes' metadata of.
        limit : int, optional
            The amount of codes you want returned. The default is 20.

        Returns
        -------
        metadata : JSON
            Detailed information surrounding the code(s).

        """
        api_call = (
            self._get_url("catalog") + "/classification/" + classification + "/codes?"
        )
        params = {}
        params.update(self._get_param(limit, "limit"))

        return _query(api_call, self.api_key, params)

    def profile(self, variable):
        """
        Gets a profile on a variable that contains statistical information.

        Parameters
        ----------
        variable : str
            The name of the variable you want the profile of.

        Returns
        -------
        profile : JSON
            Detailed information surrounding a profile on a variable.

        """
        api_call = self._get_url("catalog") + "/variables/profile?cols=" + variable

        response = get_response(api_call, self.api_key)
        return json.load(response)

    def get_metadata(self):
        """
        Gets the metadata for the dataproduct in JSON format.

        Returns
        -------
        metadata : JSON
            Detailed information surrounding the dataproduct.

        """
        api_call = self._get_url("catalog")

        response = get_response(api_call, self.api_key)
        return json.load(response)

    def _get_column_count(self, cols, collimit):
        col_count = None
        if cols == None:
            col_count_api_call = self._get_url("query") + "/select?"
            col_count_params = {}
            col_count_params.update(self._get_param(1, "limit"))

            col_count_results = _query(col_count_api_call, self.api_key, col_count_params)
            col_count = len(col_count_results["records"][0])
        else:
            col_count = len(cols)
            
        if collimit == None:
            return col_count
        else:
            return col_count if col_count < collimit else collimit

    def _get_url(self, endpoint):
        if self.catalog_id is None:
            raise ValueError("Catalog ID must be specified")

        if self.dataproduct_id is None:
            raise ValueError("Data Product ID must be specified")

        return (
            self.api
            + "/api/"
            + endpoint
            + "/"
            + self.catalog_id
            + "/"
            + self.dataproduct_id
        )

    def _get_param(self, param_values, param_name):
        if param_values is not None:
            param = ""
            if type(param_values) is list:
                value_delim = ""
                for param_value in param_values:
                    param += value_delim + str(param_value)
                    value_delim = ","
            else:
                param = str(param_values)

            return {param_name: param}
        else:
            return {}

    def _batch(self, api_call, params, max_records, limit, offset=0):
        results = []

        first_pass = True
        more_rows = True
        while (first_pass or more_rows) and (max_records == None or max_records > 0):
            first_pass = False
            api_call_copy = api_call
            params.update(self._get_param(offset, "offset"))

            if max_records == None:
                params.update(self._get_param(limit, "limit"))
                offset += limit
            else:
                if max_records > limit:
                    params.update(self._get_param(limit, "limit"))
                    offset += limit
                    max_records -= limit
                else:
                    params.update(self._get_param(max_records, "limit"))
                    max_records = 0

            result = _query(api_call_copy, self.api_key, params)
            results.append(result)

            more_rows = result["info"]["moreRows"]

        return results


class RdsResults:
    """A wrapper object that binds the records, the column names, and metadata on the columns together."""

    def __init__(self, records, columns, metadata, totals=None, count=None):
        self.records = records
        self.columns = columns
        self.metadata = metadata
        self.totals = totals
        self.count = count


def _get_metadata(results):
    metadata = {}
    for result in results:
        for variable in result["variables"]:
            var_name = ""
            try:
                var_name = variable["label"]
            except KeyError:
                var_name = variable["name"]

            metadata[var_name] = variable

    return metadata


def _query(api_call, api_key, params):
    if sys.version_info > (3, 0):
        import urllib.parse

        api_call += urllib.parse.urlencode(params)
    else:
        import urllib

        api_call += urllib.urlencode(params)

    response = get_response(api_call, api_key)
    return json.load(response)


def _get_rds_results(results, metadata_json, count):
    col_names = None
    metadata = None
    if metadata_json is not None:
        col_names = []
        for var_name in metadata_json.keys():
            col_names.append(var_name)
        metadata = list(metadata_json.values())

    records = []
    totals = None
    for result in results:
        for record in result["records"]:
            records.append(record)
        if result["totals"] != None and result["totals"] != []:
            if totals == None:
                totals = []
            for total in result["totals"]:
                totals.append(total)

    return RdsResults(records, col_names, metadata, totals, count)
